panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Service

<!--

Welcome to ingress-nginx!  For a smooth issue process, try to answer the following questions.
Don't worry if they're not all applicable; just try to include what you can :-).
More info helps better understanding of the issue (needless to say).

If you need to include code snippets or logs, please put them in fenced code
blocks.  If they're super-long, please use the details tag like
<details><summary>super-long log</summary> lots of stuff </details>

-->

<!--

IMPORTANT!!!

Please complete the next sections or the issue will be closed.
This questions are the first thing we need to know to understand the context.

-->

**What happened**:

pod panic with "interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Service".  Checked the original source code and it seems like the type assertion ![here](https://github.com/kubernetes/ingress-nginx/blob/controller-v1.4.0/internal/ingress/controller/store/store.go#L772) doesn't handle the case where the obj could be cache.DeletedFinalStateUnknown. I am sorry if this issue might have been reported by others already. 

Thanks.

<!-- (please include exact error messages if you can) -->

E0525 04:03:38.226398       7 runtime.go:79] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x17c10e0), concrete:(*runtime._type)(0x18a3e00), asserted:(*runtime._type)(0x1a23800), missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Service)
goroutine 125 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1822d40?, 0xc003232660})
        k8s.io/apimachinery@v0.25.2/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x203020203030202?})
        k8s.io/apimachinery@v0.25.2/pkg/util/runtime/runtime.go:49 +0x75
panic({0x1822d40, 0xc003232660})
        runtime/panic.go:884 +0x212
k8s.io/ingress-nginx/internal/ingress/controller/store.New.func21({0x18a3e00?, 0xc001d820e0?})
        k8s.io/ingress-nginx/internal/ingress/controller/store/store.go:772 +0xde
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
        k8s.io/client-go@v0.25.2/tools/cache/controller.go:246
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        k8s.io/client-go@v0.25.2/tools/cache/shared_informer.go:820 +0xaf
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000659f38?, {0x1ccc140, 0xc00056ea20}, 0x1, 0xc000520cc0)
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0xc000659f88?)
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:92
k8s.io/client-go/tools/cache.(*processorListener).run(0xc0007b3a80?)
        k8s.io/client-go@v0.25.2/tools/cache/shared_informer.go:812 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:75 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        k8s.io/apimachinery@v0.25.2/pkg/util/wait/wait.go:73 +0x85
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Service [recovered]
        panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Service

**What you expected to happen**:

<!-- What do you think went wrong? -->


**NGINX Ingress controller version** (exec into the pod and run nginx-ingress-controller --version.): 1.4.0
<!--
POD_NAMESPACE=ingress-nginx
POD_NAME=$(kubectl get pods -n $POD_NAMESPACE -l app.kubernetes.io/name=ingress-nginx --field-selector=status.phase=Running -o jsonpath='{.items[0].metadata.name}')

kubectl exec -it $POD_NAME -n $POD_NAMESPACE -- /nginx-ingress-controller --version  
-->

**Kubernetes version** (use `kubectl version`): 

**Environment**:

- **Cloud provider or hardware configuration**: GCP
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
  - `Please mention how/where was the cluster created like kubeadm/kops/minikube/kind etc. `
- **Basic cluster related info**:
  - `kubectl version`
  - `kubectl get nodes -o wide`

- **How was the ingress-nginx-controller installed**:
  - If helm was used then please show output of `helm ls -A | grep -i ingress`
  - If helm was used then please show output of `helm -n <ingresscontrollernamepspace> get values <helmreleasename>`
  - If helm was not used, then copy/paste the complete precise command used to install the controller, along with the flags and options used
  - if you have more than one instance of the ingress-nginx-controller installed in the same cluster, please provide details for all the instances

- **Current State of the controller**:
  - `kubectl describe ingressclasses`
  - `kubectl -n <ingresscontrollernamespace> get all -A -o wide`
  - `kubectl -n <ingresscontrollernamespace> describe po <ingresscontrollerpodname>`
  - `kubectl -n <ingresscontrollernamespace> describe svc <ingresscontrollerservicename>`

- **Current state of ingress object, if applicable**:
  - `kubectl -n <appnnamespace> get all,ing -o wide`
  - `kubectl -n <appnamespace> describe ing <ingressname>`
  - If applicable, then, your complete and exact curl/grpcurl command (redacted if required) and the reponse to the curl/grpcurl command with the -v flag

- **Others**:
  - Any other related information like ;
    - copy/paste of the snippet (if applicable)
    - `kubectl describe ...` of any custom configmap(s) created and in use
    - Any other related information that may help


**How to reproduce this issue**:
<!---

not very easy to reproduce this issue. 

As minimally and precisely as possible. Keep in mind we do not have access to your cluster or application.
Help up us (if possible) reproducing the issue using minikube or kind.

## Install minikube/kind

- Minikube https://minikube.sigs.k8s.io/docs/start/
- Kind https://kind.sigs.k8s.io/docs/user/quick-start/

## Install the ingress controller

kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/baremetal/deploy.yaml

## Install an application that will act as default backend (is just an echo app)

kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/docs/examples/http-svc.yaml

## Create an ingress (please add any additional annotation required)

echo "
  apiVersion: networking.k8s.io/v1
  kind: Ingress
  metadata:
    name: foo-bar
    annotations:
      kubernetes.io/ingress.class: nginx
  spec:
    ingressClassName: nginx # omit this if you're on controller version below 1.0.0
    rules:
    - host: foo.bar
      http:
        paths:
        - path: /
          pathType: Prefix
          backend:
            service:
              name: http-svc
              port: 
                number: 80
" | kubectl apply -f -

## make a request

POD_NAME=$(k get pods -n ingress-nginx -l app.kubernetes.io/name=ingress-nginx -o NAME)
kubectl exec -it -n ingress-nginx $POD_NAME -- curl -H 'Host: foo.bar' localhost

--->

**Anything else we need to know**:


<!-- If this is actually about documentation, uncomment the following block -->

<!-- 
/kind documentation
/remove-kind bug
-->

Maybe similar in 1.10 :

```
	k8s.io/apimachinery@v0.29.3/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x19c4120?, 0x2ced7f0?})
	runtime/panic.go:770 +0x132
k8s.io/ingress-nginx/internal/ingress/controller/store.New.func21({0x1a77ae0?, 0xc02539d5a0?})
	k8s.io/ingress-nginx/internal/ingress/controller/store/store.go:819 +0x102
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	k8s.io/client-go@v0.29.3/tools/cache/controller.go:253
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.29.3/tools/cache/shared_informer.go:977 +0x9f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000193f70, {0x1f08040, 0xc00030b680}, 0x1, 0xc0001219e0)
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00052bf70, 0x3b9aca00, 0x0, 0x1, 0xc0001219e0)
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000324ab0)
	k8s.io/client-go@v0.29.3/tools/cache/shared_informer.go:966 +0x69
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/wait.go:72 +0x52
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 41
	k8s.io/apimachinery@v0.29.3/pkg/util/wait/wait.go:70 +0x73
88.160.60.107 - - [08/Jul/2024:19:48:58 +0000] "GET /media/cache/resolve/l800lq/websites/2fc5b756a1c16945073bd0c4128313d7/img/image-leponyme-restaurant-terrasse-cannes-centre-ville-09_20220113092717.png HTTP/2.0" 302 958 "https://www.leponyme-cannes.com/le-restaurant-leponyme-bistrot-cannes" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Mobile Safari/537.36" 127 0.085 [mgnt-production-app-http] [] 10.2.18.139:80 970 0.084 302 032c5c7404185f64212eb606887e96c2
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x17013c2]
```

We experienced this issue with the 1.11.2 version. We are running 4 independent ingress controller and saw this error on 3 of this within 10 seconds.

```
E0912 11:44:11.621441       7 store.go:817] unexpected type: cache.DeletedFinalStateUnknown
E0912 11:44:11.621521       7 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 145 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1a64b60, 0x2e19780})
	k8s.io/apimachinery@v0.30.3/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0007aa700?})
	k8s.io/apimachinery@v0.30.3/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x1a64b60?, 0x2e19780?})
	runtime/panic.go:770 +0x132
k8s.io/ingress-nginx/internal/ingress/controller/store.New.func21({0x1b1ef60?, 0xc00d1eb320?})
	k8s.io/ingress-nginx/internal/ingress/controller/store/store.go:819 +0x102
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	k8s.io/client-go@v0.30.3/tools/cache/controller.go:253
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	k8s.io/client-go@v0.30.3/tools/cache/shared_informer.go:983 +0x9f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc010eb9f70, {0x1fce640, 0xc000168ff0}, 0x1, 0xc000340c00)
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000738f70, 0x3b9aca00, 0x0, 0x1, 0xc000340c00)
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000134a20)
	k8s.io/client-go@v0.30.3/tools/cache/shared_informer.go:972 +0x69
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/wait.go:72 +0x52
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 108
	k8s.io/apimachinery@v0.30.3/pkg/util/wait/wait.go:70 +0x73
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x1790242]
```

--- a/internal/ingress/controller/store/store.go
+++ b/internal/ingress/controller/store/store.go
@@ -817,7 +817,17 @@ func New(
 		DeleteFunc: func(obj interface{}) {
 			svc, ok := obj.(*corev1.Service)
 			if !ok {
-				klog.Errorf("unexpected type: %T", obj)
+				// If we reached here it means the service was deleted but its final state is unrecorded.
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					klog.ErrorS(nil, "Error obtaining object from tombstone", "key", obj)
+					return
+				}
+				svc, ok = tombstone.Obj.(*corev1.Service)
+				if !ok {
+					klog.Errorf("Tombstone contained object that is not a Service: %#v", obj)
+					return
+				}
 			}
 			if svc.Spec.Type == corev1.ServiceTypeExternalName {
 				updateCh.In() <- Event{

